informalsystems/ibc-rs
# Fix bug in event monitor where subscriptions would not be terminated properly on WebSocket error

Closes: #293

## Description

This PR fixes a potential bug where subscriptions would not be terminated properly if the WebSocket client encountered an error because

a) we were emptying the `subscriptions` vector when passing it to `select_all`, and
b) we were not closing the client explicitly

### To test this PR:

Use the following configuration file:

```toml
title = "IBC Relayer Config Example"

[global]
timeout = "10s"
strategy = "naive"

[[chains]]
  id = "chain_A"
  rpc_addr = "localhost:26657"
  account_prefix = "cosmos"
  key_name = "testkey"
  store_prefix = "ibc"
  client_ids = ["clA1", "clA2"]
  gas = 200000
  gas_adjustement = 1.3
  gas_price = "0.025stake"
  trusting_period = "336h"

[[connections]]
```

In one terminal, spawn a Tendermint node running the KV Store app:

```bash
docker run -it --rm -v "/tmp/tendermint:/tendermint" tendermint/tendermint init
docker run -it --rm -v "/tmp/tendermint:/tendermint" -p 26657:26657 tendermint/tendermint node --proxy_app=kvstore
```

In another terminal, from the `relayer-cli` folder, run the `listen` command of the relayer:

```bash
$ cargo run -- -c path/to/config/file.toml listen
```

Expected output:

```
Oct 06 18:01:35.126  INFO relayer_cli::commands::listen: spawning event monitor for chain.id=chain_A
Oct 06 18:01:35.159  INFO relayer_cli::commands::listen: spawning main event handler
Oct 06 18:01:35.160  INFO relayer::event_monitor: running listener for chain.id=chain_A
Oct 06 18:01:35.160  INFO relayer::event_handler: running IBC Event Handler
Oct 06 18:01:35.343  INFO relayer::event_handler: Chain chain_A pushed {"NewBlock":{"height":"322"}}
Oct 06 18:01:35.863  INFO relayer::event_handler: Chain chain_A pushed {"NewBlock":{"height":"323"}}

…
```

______

For contributor use:

- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
